### PR TITLE
@craigspaeth Image set

### PR DESCRIPTION
**Pull request by kanaabe**
Closes https://github.com/artsy/positron/issues/703 https://github.com/artsy/positron/issues/692 https://github.com/artsy/positron/issues/615 https://github.com/artsy/positron/issues/704

I apologies for the size of this PR! There are a couple more visual bugs I want to fix before merging but the main functionality is complete. 

image_set
Preview (4 should be centered in the svg): 
image

### CODE REVIEW

**Comment by craigspaeth** [comment]
:-o this is super cool!

File: api/apps/articles/model/schema.coffee [hunk 1]
```diff
@@ -74,20 +90,7 @@ fullscreenSection = (->
...
-          title: @string().allow('', null)
-          image: @string().allow('', null)
-          partner: @object().keys
-            name: @string().allow('', null)
-            slug: @string().allow('', null)
-          artist: @object().keys
-            name: @string().allow('', null)
-            slug: @string().allow('', null)
-      ).allow(null).default([])
+      artworks: @array().items(denormalizedArtwork).allow(null).default([])
```
Comment (craigspaeth) [inline api/apps/articles/model/schema.coffee]: :+1:

File: client/apps/edit/components/section_image_set/index.coffee [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        $(@refs.byUrl.getDOMNode()).val ''
+        @state.images.push artwork.denormalized()
+        @onStateChange()
+
+  showPreviewImages: ->
+    allowedPixels = 500 - 40.0
+    totalPixels = 0.0
+    usedPixels = 0.0
+    $('.esis-preview-container .esis-preview-image').each (i, value) ->
+      _.defer ->
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/index.coffee]: Good ol defer to the resuce.

File: client/apps/edit/components/section_image_set/index.coffee [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+React = require 'react'
+toggleScribePlaceholder = require '../../lib/toggle_scribe_placeholder.coffee'
+sd = require('sharify').data
+icons = -> require('./icons.jade') arguments...
+Autocomplete = require '../../../../components/autocomplete/index.coffee'
+Artwork = require '../../../../models/artwork.coffee'
+{ div, section, h1, h2, span, img, header, input, nav, a, button, p, ul, li, strong } = React.DOM
+{ crop, resize, fill } = embedly = require('embedly-view-helpers')(sd.EMBEDLY_KEY)
+
+module.exports = React.createClass
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/index.coffee]: :clap: Nice work on this component! It's a little long in line numbers, but not _too_ long yet. If this grows in size I'd encourage trying to find way to break it down into even smaller parts.

File: client/apps/edit/components/section_image_set/index.coffee [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            </div>
+            #{data.value}
+          """
+      selected: @onSelect
+    _.defer -> $el.focus()
+
+  onSelect: (e, selected) ->
+    new Artwork(id: selected.id).fetch
+      success: (artwork) =>
+        @state.images.push artwork.denormalized()
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/index.coffee]: You shouldn't be mutating state directly here. Make sure to always use `setState` instead. e.g.

````coffeescript
newImages = @state.images.concat [artwork.denormalized()]
@setState images: newImages
````

File: client/apps/edit/components/section_image_set/index.coffee [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+      key: sd.GEMINI_KEY
+      progress: (percent) =>
+        @setState progress: percent
+      add: (src) =>
+        @setState progress: 0.1
+      done: (src) =>
+        image = new Image()
+        image.src = src
+        image.onload = =>
+          @state.images.push url: src, type: 'image'
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/index.coffee]: Ditto as above

File: client/apps/edit/components/section_image_set/index.coffee [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+          @setState progress: null
+          @onStateChange()
+
+  removeItem: (item) -> =>
+    @state.images = _.without @state.images, item
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/index.coffee]: :eyes:

File: client/apps/edit/components/section_image_set/index.coffee [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    new Artwork(id: slug).fetch
+      error: (m, res) =>
+        @refs.byUrl.setState(
+          errorMessage: 'Artwork not found. Make sure your urls are correct.'
+          loadingUrls: false
+        ) if res.status is 404
+      success: (artwork) =>
+        @refs.byUrl.setState loading: false, errorMessage: ''
+        $(@refs.byUrl.getDOMNode()).val ''
+        @state.images.push artwork.denormalized()
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/index.coffee]: :eye

File: client/apps/edit/components/section_image_set/index.coffee [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+  componentDidUpdate: ->
+    @attachScribe()
+
+  componentWillUnmount: ->
+    @autocomplete.remove()
+
+  onStateChange: ->
+    @props.section.set images: @state.images if @state.images.length > 0
+    @showPreviewImages()
+    @forceUpdate()
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/index.coffee]: You also should have to use `forceUpdate` unless you're hooking into something un-react-like (e.g. events on Backbone models). I bet if you refactor the state mutations into `setState` you'll notice you won't need this.

Also if `onStateChange` is meant to run after every "re-render" than I think a [lifecycle callback](https://facebook.github.io/react/docs/component-specs.html#updating-componentdidupdate) would be more appropriate.

File: client/apps/edit/components/section_image_set/index.styl [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+    &::after
+      top 3px !important
+    svg
+      top 7px !important
+
+.esis-nav, .esis-caption
+  height 40px
+  display inline-block
+
+[data-useragent*=Firefox] .esis-nav
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/index.styl]: :+1:

File: client/apps/edit/components/section_image_set/index.styl [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+artwork-img-bottom = 10px
+short-input-width = 290px
+wide-input-width = 640px
+
+.edit-section-image-set
+  position relative
+  margin-top 20px
+
+.edit-section-image-set .edit-section-controls
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/index.styl]: Great namespacing/stylus work! Likewise with the line numbers here—not _too_ long yet, but if this thing grows also consider breaking down the stylesheets.

File: client/apps/edit/components/section_image_set/test/index.coffee [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+React = require 'react'
+require 'react/addons'
+r =
+  find: React.addons.TestUtils.findRenderedDOMComponentWithClass
+  simulate: React.addons.TestUtils.Simulate
+  findAll: React.addons.TestUtils.scryRenderedDOMComponentsWithClass
+{ div } = React.DOM
+fixtures = require '../../../../../../test/helpers/fixtures'
+
+describe 'SectionImageSet', ->
```
Comment (craigspaeth) [inline client/apps/edit/components/section_image_set/test/index.coffee]: :heart: the tests!

File: client/models/artwork.coffee [hunk 12]
```diff
@@ -17,3 +17,39 @@ module.exports = class Artwork extends Backbone.Model
 
   defaultImage: ->
     new AdditionalImages(@get('images')).default()
+
+  denormalized: ->
```
Comment (craigspaeth) [inline client/models/artwork.coffee]: :+1: :+1:

File: client/models/artwork.coffee [hunk 13]
```diff
@@ -17,3 +17,39 @@ module.exports = class Artwork extends Backbone.Model
...
+    title: @get('title')
+    image: @defaultImage().bestImageUrl(['larger','large', 'medium', 'small'])
+    partner:
+      name: @getPartnerName()
+      slug: @getPartnerLink()
+    artist:
+      name: @getArtistName()
+      slug: @get('artist')?.id
+
+  getArtistName: ->
```
Comment (craigspaeth) [inline client/models/artwork.coffee]: :+1:

Although these might be something we want to [share across Backbone apps](https://github.com/artsy/artsy-backbone-mixins/tree/master/lib), or leverage Metaphysics for. Totally cool with this implementation for now though.

**Comment by craigspaeth** [comment]
Looking good! Mainly concerned about the `state` mutations.

Also might want to consider moving the denormalizing artwork logic to the api-side and just have the client send down artwork ids. You could probably leverage some of the api-side code you wrote before in this case, and it would make it friendly for Orta when he writes a native iPad Writer app :stuck_out_tongue:

**Comment by kanaabe** [comment]
Thanks for the comments @craigspaeth!

**Comment by kanaabe** [comment]
I think the biggest opportunity after this PR is to make the image and artwork components pluggable into the image_set component. It's admittedly a lot of repeated code in various places. Anyway, I think this one is good to go now, I've changed the mutations and hooked into the `componentWillUpdate` lifecycle cb and removed `forceUpdate()` (you're right we totally didn't need this!)

**Comment by craigspaeth** [comment]
w00t w00t! Nice work! :confetti_ball:

**Pull request merged**